File: src/etch/bindings/csharp/support/Class2TypeMap.cs
using System;
using System.Collections.Generic;
using System.Text;
using Etch.Msg;

namespace Etch.Support
{
    public class Class2TypeMap
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="c"></param>
        /// <returns>the Type for the specified class, or null</returns>
        public XType Get( Type c )
        {
            try
            {
                return c2t[ c ];
            }
            catch ( KeyNotFoundException )
            {
                return null;
            }
        }

        /// <summary>
        /// Adds a map entry from c to t
        /// </summary>
        /// <param name="c"></param>
        /// <param name="t"></param>
        public void Add( Type c, XType t )
        {
            if ( locked )
                throw new Exception( "locked" );

            XType x;
            try
            {
                x = c2t[ c ];
            }
            catch ( KeyNotFoundException )
            {
                x = null;
            }

            if ( x != null )
            {
                if ( !x.Equals( t ) )
                    throw new ArgumentException( String.Format( "type {0} : class {1} is already mapped to type {2}", t, c, x ) );
                return;
            }
            c2t.Add( c, t );
        }

        /// <summary>
        /// Adds all the mappings from other to this
        /// </summary>
        /// <param name="other"></param>
        public void AddAll( Class2TypeMap other )
        {
            if ( locked )
                throw new Exception( "locked" );

            foreach ( KeyValuePair<Type, XType> me in other.c2t )
            {
                Add( me.Key, me.Value );
            }
        }

        /// <summary>
        /// Locks the map, disallowing any more changes
        /// </summary>
        public void Lock()
        {
            locked = true;
        }

        private readonly Dictionary<Type, XType> c2t = new Dictionary<Type, XType>();

        private Boolean locked;
    }
}

File: src/etch/bindings/java/support/Class2TypeMap.java
package etch.bindings.java.support;

import java.util.HashMap;
import java.util.Map;

import etch.bindings.java.msg.Type;

/**
 * Mapping between Class and Type.
 */
public class Class2TypeMap
{
	/**
	 * @param c
	 * @return the Type for the specified class, or null.
	 */
	public Type get( Class<?> c )
	{
		return c2t.get( c );
	}

	/**
	 * Adds a map entry from c to t.
	 * @param c
	 * @param t
	 * @throws IllegalStateException if locked.
	 * @throws IllegalArgumentException if conflict.
	 */
	public void put( Class<?> c, Type t )
	{
		if (locked)
			throw new IllegalStateException( "locked" );
		
		Type x = c2t.get( c );
		if (x != null)
		{
			if (!x.equals( t ))
				throw new IllegalArgumentException( String.format(
					"type %s: class %s is already mapped to type %s",
					t, c, x ) );
			
			// x.equals( t ) therefore our work is finished.
			return;
		}
		
		c2t.put( c, t );
	}

	/**
	 * Adds all the mappings from other to this.
	 * @param other
	 * @throws IllegalStateException if locked.
	 * @throws IllegalArgumentException if conflict.
	 */
	public void addAll( Class2TypeMap other )
	{
		if (locked)
			throw new IllegalStateException( "locked" );
		
		for (Map.Entry<Class<?>, Type> me: other.c2t.entrySet())
			put( me.getKey(), me.getValue() );
	}

	/**
	 * Locks the map, disallowing any more changes.
	 */
	public void lock()
	{
		locked = true;
	}
	
	private final Map<Class<?>, Type> c2t = new HashMap<Class<?>, Type>();
	
	private boolean locked;
}